### PR TITLE
Drag and Drop

### PR DESCRIPTION
**Pull request by mruwnik**

### CODE REVIEW

File: app/hooks/stateModifiers.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import {Question, QuestionState, RelatedQuestions, PageId} from '~/server-utils/stampy'
```
Comment (mruwnik) [inline app/hooks/stateModifiers.tsx:1]: these are mainly extracted from `useQuestionStateInUrl`

File: app/hooks/stateModifiers.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  return `${pageid}-${state.replace(removePageRe, '')}`
+}
+
+/*
+ * Add all the provided `questions` to the state string as collapsed
+ */
+export const addQuestions = (state: StateString, questions: Question[]): StateString =>
+  questions.reduce((newState, q) => `${newState}${q.pageid}${QuestionState.COLLAPSED}`, state)
+
+export const insertAfter = (state: StateString, pageId: PageId, to: PageId): StateString => {
```
Comment (mruwnik) [inline app/hooks/stateModifiers.tsx:25]: this function is new

File: app/hooks/stateModifiers.tsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    })
+    .join('')
+
+export const removeQuestion = (state: StateString, pageid: PageId): StateString =>
+  getStateEntries(state)
+    .filter((q) => q[0] != pageid)
+    .flat()
+    .join('')
+
+export const moveQuestion = (state: StateString, pageid: PageId, to: PageId): StateString => {
```
Comment (mruwnik) [inline app/hooks/stateModifiers.tsx:75]: as is this one

File: app/routes/index.tsx [hunk 4]
```diff
@@ -32,6 +33,52 @@ export const unstable_shouldReload: ShouldReloadFunction = () => false
 
 const year = new Date().getFullYear()
 
+const Header = () => {
```
Comment (mruwnik) [inline app/routes/index.tsx:36]: this was annoying me, as I had to keep scrolling up, so I extracted it

File: app/routes/index.tsx [hunk 5]
```diff
@@ -82,66 +130,49 @@ export default function App() {
...
+          onDragOver={handleDragOver()}
+        >
           {questions.map((question) => (
-            <Question
-              key={question.pageid}
-              questionProps={question}
-              onLazyLoadQuestion={onLazyLoadQuestion}
-              onToggle={toggleQuestion}
-            />
+            <Fragment key={question.pageid}>
```
Comment (mruwnik) [inline app/routes/index.tsx:158]: I considered having an additional element to wrap this and the drag placeholder, but it kept eating my event handlers, so I decided this was good enough

File: app/routes/questions/$question.tsx [hunk 6]
```diff
@@ -33,11 +33,13 @@ export function Question({
   onLazyLoadQuestion,
   onToggle,
   selectQuestion,
+  ...props
 }: {
   questionProps: Question
   onLazyLoadQuestion: (question: Question) => void
   onToggle: ReturnType<typeof useQuestionStateInUrl>['toggleQuestion']
   selectQuestion: (pageid: string, title: string) => void
+  [k: string]: JSX.IntrinsicElements['div']
```
Comment (Aprillion) [inline app/routes/questions/$question.tsx:42]: this is used for `article` element,, but I guess there is no difference in type of props, so probably fine to keep

File: app/routes/index.tsx (hunk 5 above)
Comment (Aprillion) [reply app/routes/index.tsx:158]: cool

File: app/hooks/useDraggable.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    e.dataTransfer.setData('text', question.title)
+  }
+
+  const handleDragEnd = (question: Question) => () => {
+    onMove(question.pageid, draggedOver)
+    setDraggedOver(null)
+  }
+
+  const DragPlaceholder = ({pageid}: {pageid: PageId}) => {
+    if (draggedOver == pageid) return <hr className="drag-placeholder"></hr>
```
Comment (Aprillion) [inline app/hooks/useDraggable.tsx:48]: I thought prettier would change this to `<hr className="drag-placeholder" />` - if it doesn't happen automatically, I don't mind either way, but could you please double check that prettier is happy?

File: app/hooks/stateModifiers.tsx (hunk 3 above)
Comment (Aprillion) [reply app/hooks/stateModifiers.tsx:75]: looks good, but please keep including issue numbers for context, in this case #52, it's easier to understand some code when I check what features have been asked for, e.g. `removeQuestion` sounded weird without context 😅

**Review by Aprillion — APPROVED** [review]

File: app/hooks/useDraggable.tsx (hunk 7 above)
Comment (mruwnik) [reply app/hooks/useDraggable.tsx:48]: it's definately happy, as this isn't how it looked when I wrote it :D

**Pull request merged**